biogtr/io/association_matrix.py

```diff
@@ -0,0 +1,302 @@
+                        Otherwise label by the query_instance indices
+            col_label: How to label the columns(references).
+                        If `gt` then label by gt track id.
+                        If `pred` then label by pred track id.
+                        Otherwise label by the ref_instance indices
+
+        Returns:
+            The association matrix as a pandas dataframe.
+        """
+        matrix = self.numpy()
+
+        if row_label.lower() == "gt":
+            row_inds = [
+                instance.gt_track_id.item() for instance in self.query_instances
+            ]
+
+        elif row_label.lower() == "pred":
+            row_inds = [
+                instance.pred_track_id.item() for instance in self.query_instances
+            ]
+
+        else:
+            row_inds = np.arange(len(self.query_instances))
+
+        if col_label.lower() == "gt":
+            col_inds = [instance.gt_track_id.item() for instance in self.ref_instances]
+
+        elif col_label.lower() == "pred":
+            col_inds = [
+                instance.pred_track_id.item() for instance in self.ref_instances
+            ]
+
+        else:
+            col_inds = np.arange(len(self.ref_instances))
+
+        asso_df = pd.DataFrame(matrix, index=row_inds, columns=col_inds)
+
+        return asso_df
+
+    def reduce(
+        self,
+        to: tuple[str] = ("inst", "traj"),
+        by: tuple[str] = (None, "pred"),
+        reduce_method: callable = np.sum,
+    ) -> pd.DataFrame:
+        """Reduce association matrix rows/columns to inst/traj x traj/inst.
+
+        Args:
+           to: A tuple indicating how to reduce rows/columns. Either inst (remains unchanged), or traj (reduces matrix from n_query/n_ref to n_traj
+           by: A tuple indicating how to group rows/columns when aggregating. Either "pred" or "gt"
+           method: A callable function that operates on numpy matrices and can take an `axis` arg for reducing.
+
+        Returns:
+            The association matrix reduced to an inst/traj x traj/inst association matrix as a dataframe.
+        """
+        row_to, col_to = to
+        row_by, col_by = by
+
+        if row_to not in ("inst", "traj") or col_to not in ("inst", "traj"):
+            raise ValueError(
+                f"Can only reduce to inst/traj x inst/traj but ({row_to} x {col_to}) was requested!"
+            )
+
+        if row_by not in ("pred", "gt", None) or col_by not in ("pred", "gt", None):
+            raise ValueError(
+                f"Can aggregate by [gt, pred, None] but {row_by} and {col_by} was requested!"
+            )
+
+        n_rows = len(self.query_instances)
+        n_cols = len(self.ref_instances)
+
+        col_tracks = {-1: self.ref_instances}
+        row_tracks = {-1: self.query_instances}
+
+        col_inds = [i for i in range(len(self.ref_instances))]
+        row_inds = [i for i in range(len(self.query_instances))]
+
+        if "tra" in col_to:
```

Comment:
Too clever bud. What if I provide `"instrance"`???

Be explicit and list out the conditions with equality checks. `if col_to == "track" or col_to == "trajectory"`